File: Stacks/FinalPriceWithSpecialDisc/final.cpp
#include <vector>
#include <stack>
using namespace std;

class Solution {
public:
    vector<int> finalPrices(vector<int>& arr) {
        int n = arr.size();
        vector<int>ans(n);
        ans[n-1]=arr[n-1];
        stack<int>st;
        st.push(arr[n-1]);
        for(int i=n-2;i>=0;i--){
            while(st.size() && st.top()>arr[i])st.pop();
            if(st.size())ans[i]=arr[i]-st.top();
            else ans[i]=arr[i];
            st.push(arr[i]);
        }
        return ans;
    }
};

File: Stacks/FinalPriceWithSpecialDisc/final.py
from typing import List

class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        n = len(prices)
        st = []
        ans = prices[:]
        st.append(prices[n-1])
        for i in range(n-2,-1,-1):
            while len(st) and prices[i]<st[-1]: st.pop()
            if len(st): ans[i]=prices[i]-st[-1]
            st.append(prices[i])
        return ans